### PR TITLE
Add hotkey to code block button

### PR DESCRIPTION
**Pull request by aaron-tdouble**
## Description
This PR does the following:

* adds a hotkey of `C` for the `MarkdownCodeButtonElement`
* forces hotkeys to be case sensitive

## Issue
References: https://github.com/github/special-projects/issues/135

## Screenshots
Demonstrating the current behavior of hotkeys not being case sensitive:

Screen Cast 2021-03-18 at 1 16 40 PM

## Risk Assessment
**Low risk**

### CODE REVIEW

**Comment by relph** [comment]
> adds a hotkey of C for the MarkdownCodeButtonElement

@aaron-tdouble in chrome the shortcut `cmd+shift+C` opens dev tools. Will this conflict with that? Or when you're writing MD and focused in the text field will `cmd+shift+C` insert the code block and when not focused it'll continue to open dev tools? 

Also, can we add the new shortcut to the tooltip that appears when hovering on the code icon in the markdown toolbar? Similar to bold: 

Screen Shot 2021-03-18 at 11 35 46 AM

**Comment by aaron-tdouble** [comment]
> in chrome the shortcut cmd+shift+C opens dev tools. Will this conflict with that? Or when you're writing MD and focused in the text field will cmd+shift+C insert the code block and when not focused it'll continue to open dev tools?

@relph Correct!  Because it is intercepting the event in the text area, it does not open dev tools.

> Also, can we add the new shortcut to the tooltip that appears when hovering on the code icon in the markdown toolbar? 

Absolutely!  That change will happen inside the monorepo.

File: src/index.ts [hunk 1]
```diff
@@ -389,7 +394,8 @@ function findHotkey(toolbar: Element, key: string): HTMLElement | null {
 
 function shortcut(toolbar: Element, event: KeyboardEvent) {
   if ((event.metaKey && modifierKey === 'Meta') || (event.ctrlKey && modifierKey === 'Control')) {
-    const button = findHotkey(toolbar, event.key)
+    const key = event.shiftKey ? event.key.toUpperCase() : event.key
+    const button = findHotkey(toolbar, key)
```
Comment (koddsson) [inline src/index.ts:398]: I don't think this change is needed. I can revert this block locally and the tests still run green.

The case-sensitivity is already backed in since we do strict equality of the key pressed and the `hotkey` attribute.

**Comment by koddsson** [comment]
> > in chrome the shortcut cmd+shift+C opens dev tools. Will this conflict with that? Or when you're writing MD and focused in the text field will cmd+shift+C insert the code block and when not focused it'll continue to open dev tools?
> 
> @relph Correct! Because it is intercepting the event in the text area, it does not open dev tools.

I never use this shortcut but this feels like this is unexpected behavior. I wonder if users will get upset if we override their browser shortcuts.

File: src/index.ts (hunk 1 above)
Comment (aaron-tdouble) [reply src/index.ts:398]: @koddsson I must be doing something wrong because this test fails for me on `main`.

```
code
        ✔ surrounds a line with backticks if you click the code icon
        ✖ surrounds a line with backticks via hotkey
        ✔ surrounds multiple lines with triple backticks if you click the code icon
        ✔ removes backticks from a line if you click the code icon again
        ✔ removes triple backticks on multiple lines if you click the code icon
      links
        ✔ inserts link syntax with cursor in description
        ✔ selected url is wrapped in link syntax with cursor in description
        ✔ cursor on url is wrapped in link syntax with cursor in description
        ✔ selected plan text is wrapped in link syntax with cursor in url
      images
        ✔ inserts image syntax with cursor in description
        ✔ selected url is wrapped in image syntax with cursor in description
        ✔ cursor on url is wrapped in image syntax with cursor in description
        ✔ selected plan text is wrapped in image syntax with cursor in url
      header
        ✔ inserts header syntax with cursor in description
        ✔ inserts header 1 syntax with cursor in description
        ✔ does not insert header for invalid level

Finished in 0.013 secs / 0.091 secs @ 22:03:40 GMT-0400 (Eastern Daylight Time)

SUMMARY:
✔ 86 tests completed
✖ 1 test failed

FAILED TESTS:
  markdown-toolbar-element
    after tree insertion
      code
        ✖ surrounds a line with backticks via hotkey
          Chrome Headless 89.0.4389.90 (Mac OS 11.1.0)
        AssertionError: expected '`|puts \'Hello, world!\'|`' to equal '|puts \'Hello, world!\'|'
            at Context.<anonymous> (test.js:621:16)
```

Comment (aaron-tdouble) [reply src/index.ts:398]: What does it do for you on the [example page](https://github.github.com/markdown-toolbar-element/examples/)?

If I `cmd-b` or `cmd-B`, it bolds the text.  This is true for me on Chrome, Safari, and Firefox.

Comment (aaron-tdouble) [reply src/index.ts:398]: @koddsson Wait.  Scratch my first comment above.  I was testing the wrong thing.  Let me look into that in the morning, but my second comment ☝️ still stands.  I'm curious how this behaves for you.

Comment (koddsson) [reply src/index.ts:398]: > What does it do for you on the [example page](https://github.github.com/markdown-toolbar-element/examples/)?
> 
> If I `cmd-b` or `cmd-B` (cmd-shift-b), it bolds the text. I would expect it only to respond to a lowercase `b`.

<kbd>⌘</kbd>+<kbd>B</kbd> is a global shortcut on my mac that launches a application 😄 

But yeah if I turn that off then I can bold with both. This is unexpected but apparently, if you are holding the meta key down `KeyboardEvent.key` will always be lowercase.

I made a little site to test out the `keydown` handler [here](https://js-is.glitch.me/test/uppercase-key-when-holding-metakey.html).

Getting the tests to fail when this block is removed would be good though 👍🏻

Comment (keithamus) [reply src/index.ts:398]: Oh yeah... this is a specific edge case with macos reporting lower case keys for specifically `cmd+shift`. See https://bugs.webkit.org/show_bug.cgi?id=174782 & https://bugs.chromium.org/p/chromium/issues/detail?id=747358

Comment (aaron-tdouble) [reply src/index.ts:398]: Good finds and information, @koddsson and @keithamus.  I'll work on getting a breaking test on this.

Comment (aaron-tdouble) [reply src/index.ts:398]: - [X] add [breaking test](https://github.com/github/markdown-toolbar-element/pull/37/commits/5c1807c8194563d91044b57e3ae5acb63c586dec) for lower/uppercase hotkey

**Comment by aaron-tdouble** [comment]
> I never use this shortcut but this feels like this is unexpected behavior. I wonder if users will get upset if we override their browser shortcuts.

I think this is a good question to raise.  @relph do you want to weigh in here?

On one hand, I have all kinds of tools installed that override some default shortcuts.  (In fact, I had to install [ShortcutDetective](https://www.irradiatedsoftware.com/labs/) just to figure out why `cmd-C` wasn't initially working for me in this PR.  It was due to [ColorSlurp](https://colorslurp.com/).)

The difference is that I **chose** to install those apps, even if I didn't know it was going to hijack those shortcuts.  In this case, we are in some ways forcing the shortcut on the user.

On the other hand, this (1) is very isolated to having focus in the textarea element and (2) seems to be a common keystroke for insertion of code blocks.  For some users, this will be a net good thing™.

Thoughts, opinions?

**Comment by koddsson** [comment]
> On the other hand, this (1) is very isolated to having focus in the textarea element and (2) seems to be a common keystroke for insertion of code blocks. For some users, this will be a net good thing™.

I buy that argument. Finding a shortcut that isn't used by any other application is tricky.

If we could somehow avoid this one since it's used a lot by developers and they make up the majority of our users I think that could be great but I'll default to you here.

**Comment by relph** [comment]
> On the other hand, this (1) is very isolated to having focus in the textarea element and (2) seems to be a common keystroke for insertion of code blocks. For some users, this will be a net good thing™.

I was going to argue for this ☝️ but Chrome being by far the most popular browser used by developers I have a felling we might be adding friction to a lot of peoples workflow. Feels easily avoidable. What if instead we follow notions lead and use `⌘+E`

Screen Shot 2021-03-19 at 8 34 58 AM

Thoughts?

**Comment by aaron-tdouble** [comment]
@relph I'm good with `⌘+E` and since we will have toolbar hover help, there will be some user nudging.

**Comment by aaron-tdouble** [comment]
@koddsson @keithamus I changed the keyboard shortcut to ⌘+E for code blocks.  Let me know if you have any other concerns or need anything else in order to merge this.  Thanks!

**Review by koddsson — APPROVED** [review]
This looks good to me! ✨

**Review by keithamus — APPROVED** [review]
Agreed! :+1:
